### PR TITLE
Fix cron calculation bug with @every expression

### PR DESCRIPTION
**Pull request by yux0**
Fix cron calculation bug with @every expression.
#2141

### CODE REVIEW

File: service/history/historyEngine.go [hunk 1]
```diff
@@ -766,7 +765,12 @@ func (e *historyEngineImpl) DescribeWorkflowExecution(
 	if executionInfo.HasRetryPolicy && (executionInfo.Attempt > 0) {
```
Comment (yycptt) [inline service/history/historyEngine.go]: You can also delete this part. The FirstDecisionTaskBackoffSeconds covers both retry and cron.

File: service/history/stateBuilder.go [hunk 2]
```diff
@@ -658,7 +658,7 @@ func (b *stateBuilderImpl) scheduleWorkflowTimerTask(event *shared.HistoryEvent,
 	timeout := now.Add(time.Duration(msBuilder.GetExecutionInfo().WorkflowTimeout) * time.Second)
 
 	cronSchedule := b.msBuilder.GetExecutionInfo().CronSchedule
-	cronBackoffDuration := backoff.GetBackoffForNextSchedule(cronSchedule, now)
+	cronBackoffDuration := backoff.GetBackoffForNextSchedule(cronSchedule, now, now)
```
Comment (yycptt) [inline service/history/stateBuilder.go]: Let's talk offline about this. I am not sure this is correct.

File: common/backoff/cron.go [hunk 3]
```diff
@@ -54,16 +54,24 @@ func GetBackoffForNextSchedule(cronSchedule string, nowTime time.Time) time.Dura
 		return NoBackoff
 	}
 
-	nowTime = nowTime.In(time.UTC)
-	backoffInterval := schedule.Next(nowTime).Sub(nowTime)
+	startUTCTime := startTime.In(time.UTC)
+	scheduleAfterUTCTime := scheduleAfterTime.In(time.UTC)
+	if schedule.Next(startUTCTime).Before(scheduleAfterUTCTime) {
+		nextWaitDuration := schedule.Next(startUTCTime).Sub(startUTCTime)
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/backoff/cron.go]: done

File: common/backoff/cron.go [hunk 4]
```diff
@@ -43,8 +43,8 @@ func ValidateSchedule(cronSchedule string) error {
 }
 
 // GetBackoffForNextSchedule calculates the backoff time for the next run given
-// a cronSchedule and current time
-func GetBackoffForNextSchedule(cronSchedule string, nowTime time.Time) time.Duration {
+// a cronSchedule, workflow start time and workflow close time
+func GetBackoffForNextSchedule(cronSchedule string, startTime time.Time, scheduleAfterTime time.Time) time.Duration {
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/backoff/cron.go]: done

File: service/history/historyEngine.go (hunk 1 above)
Comment (yux0) [reply service/history/historyEngine.go]: done

File: common/backoff/cron.go [hunk 5]
```diff
@@ -54,16 +54,24 @@ func GetBackoffForNextSchedule(cronSchedule string, nowTime time.Time) time.Dura
...
 	}
 
-	nowTime = nowTime.In(time.UTC)
-	backoffInterval := schedule.Next(nowTime).Sub(nowTime)
+	startUTCTime := startTime.In(time.UTC)
+	scheduleAfterUTCTime := scheduleAfterTime.In(time.UTC)
+	if schedule.Next(startUTCTime).Before(scheduleAfterUTCTime) {
+		nextWaitDuration := schedule.Next(startUTCTime).Sub(startUTCTime)
+		startToEndDuration := scheduleAfterUTCTime.Sub(startUTCTime)
+		multiplier := startToEndDuration / nextWaitDuration
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/backoff/cron.go]: Done

File: service/history/stateBuilder.go (hunk 2 above)
Comment (yux0) [reply service/history/stateBuilder.go]: Changed to get this from start event

File: service/history/stateBuilder.go [hunk 6]
```diff
@@ -658,12 +658,22 @@ func (b *stateBuilderImpl) scheduleWorkflowTimerTask(event *shared.HistoryEvent,
...
+	if ok &&
+		startEvent.WorkflowExecutionStartedEventAttributes != nil &&
+		startEvent.WorkflowExecutionStartedEventAttributes.FirstDecisionTaskBackoffSeconds != nil {
+		backoffDuration = time.Duration(*startEvent.WorkflowExecutionStartedEventAttributes.FirstDecisionTaskBackoffSeconds) * time.Second
+	}
+	if backoffDuration != backoff.NoBackoff {
+		timeout = timeout.Add(backoffDuration)
+		timeoutType := persistence.WorkflowBackoffTimeoutTypeRetry
+		if len(cronSchedule) > 0 {
+			timeoutType = persistence.WorkflowBackoffTimeoutTypeCron
```
Comment (yycptt) [inline service/history/stateBuilder.go]: This is kind of interesting... I don't kind you can determine the type just by looking cronSchedule. For example, a cron workflow with retry policy and needs retry.

Actually, you fixed another bug. Seems like previously we won't create this backoff timer at passive side if workflow needs retry. Maybe do some local verification or add a xdc integration test?

File: service/history/stateBuilder.go [hunk 7]
```diff
@@ -658,12 +658,22 @@ func (b *stateBuilderImpl) scheduleWorkflowTimerTask(event *shared.HistoryEvent,
...
 
 	cronSchedule := b.msBuilder.GetExecutionInfo().CronSchedule
-	cronBackoffDuration := backoff.GetBackoffForNextSchedule(cronSchedule, now)
-	if cronBackoffDuration != backoff.NoBackoff {
-		timeout = timeout.Add(cronBackoffDuration)
+	backoffDuration := backoff.NoBackoff
+	startEvent, ok := msBuilder.GetStartEvent()
+	if ok &&
+		startEvent.WorkflowExecutionStartedEventAttributes != nil &&
+		startEvent.WorkflowExecutionStartedEventAttributes.FirstDecisionTaskBackoffSeconds != nil {
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/stateBuilder.go]: Actually, if this field is not nil, but is 0. the backoffDuration won't equal to backoff.NoBackoff and we end up creating a useless backoff timer.

File: common/backoff/cron.go [hunk 8]
```diff
@@ -53,17 +53,21 @@ func GetBackoffForNextSchedule(cronSchedule string, nowTime time.Time) time.Dura
...
 		return NoBackoff
 	}
-
-	nowTime = nowTime.In(time.UTC)
-	backoffInterval := schedule.Next(nowTime).Sub(nowTime)
+	startUTCTime := startTime.In(time.UTC)
+	closeUTCTime := closeTime.In(time.UTC)
+	//Calculate the next schedule start time which is nearest to the close time
+	for schedule.Next(startUTCTime).Before(closeUTCTime) {
+		startUTCTime = schedule.Next(startUTCTime)
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/backoff/cron.go]: Yes. You are right

File: service/history/historyEngine.go [hunk 9]
```diff
@@ -763,10 +761,11 @@ func (e *historyEngineImpl) DescribeWorkflowExecution(
...
 	// each time DescribeWorkflowExecution is called.
 	backoffDuration := time.Duration(0)
-	if executionInfo.HasRetryPolicy && (executionInfo.Attempt > 0) {
-		backoffDuration = time.Duration(float64(executionInfo.InitialInterval)*math.Pow(executionInfo.BackoffCoefficient, float64(executionInfo.Attempt-1))) * time.Second
-	} else if len(executionInfo.CronSchedule) != 0 {
-		backoffDuration = backoff.GetBackoffForNextSchedule(executionInfo.CronSchedule, executionInfo.StartTimestamp)
+	startEvent, ok := msBuilder.GetStartEvent()
+	if ok &&
+		startEvent.WorkflowExecutionStartedEventAttributes != nil &&
+		startEvent.WorkflowExecutionStartedEventAttributes.FirstDecisionTaskBackoffSeconds != nil {
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/historyEngine.go]: done

File: service/history/stateBuilder.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/stateBuilder.go]: done

File: service/history/stateBuilder.go (hunk 6 above)
Comment (yux0) [reply service/history/stateBuilder.go]: done

**Pull request closed without merging**